### ISSUE TITLE
Reducing bloat in build process by removing bundling

### ISSUE THREAD

**Issue by dillonkearns** [dillonkearns/elm-pages#148]
As part of the no-webpack build (#140), I'm removing some bloat from `elm-pages`, and as a result having it be more focused on pre-rendering and doing things with Elm code, and doing very minimal steps with JS/CSS/assets.

I described this philosophy in this comment earlier https://github.com/dillonkearns/elm-pages/issues/70#issuecomment-596606802. I've gotten more clarity around this so I'm stating my clarified vision here so people can comment and share feedback.

## Goal
The overall goal is that the lowest common denominator is fast performance and a non-bloated build step. If elm-pages takes on too much in terms of bundling, image optimization, etc., then it will require every user to take on that as the minimum bloat, including:

- Bundling JS step
- Bundling/processing CSS
- Building up the dependency graph of these things through webpack (or any bundler)
- Doing an image optimization, or at the very least installing those dependencies even if a user opts out of using them in their build

I'm seeing more and more that this leads to bloated builds, and also a more complex and error-prone build step. So the goal is to have elm-pages be responsible for "just making things work" with Elm code and site configuration in pure Elm. But things outside of Elm are very minimal.

## Minimal non-Elm processes
What does that minimal process look like?

### No JavaScript Bundling

Modern browsers have had ES module imports working natively as a stable feature for a while now. So we can not only import JavaScript code, but we can do dynamic imports, and all without a bundler tool. If the user wants to do bundling, they could introduce a bundling step separate from the elm-pages build. If a user doesn't need a bundling step (many elm-pages sites will have very minimal JS), then they benefit from having a very fast build that doesn't do any bundling.

The current design I'm working on for this is just to have a single JavaScript entrypoint for elm-pages. elm-pages will take care of loading that JS, and giving you a Promise that you can use to add in ports to the Elm app, or run JS before/after the Elm app is hydrated. If you want to do a bundling step in user-land, you just bundle to the entrypoint file as the output.

### No CSS Bundling or Pre-Processing
Similar to the JS approach, the current design I'm working on gives you a single CSS entry point. CSS has an [import syntax](https://developer.mozilla.org/en-US/docs/Web/CSS/@import) that browsers support natively. Many elm-pages sites will do most of their CSS through Elm directly (elm-ui, elm-css, inline style tags, etc.). So for these users, there is no cost of a bundling process, pre-processors, building up a dependency graph, etc.

For users that have complex CSS pre-processing, just as with the JS approach, they can use whatever pre-processing and bundling steps they would like to use, and then output the resulting bundled/pre-processed CSS to the single entrypoint CSS file that elm-pages provides.

### No image processing
I'm realizing more and more that doing image optimization during the build step slows builds down significantly, and may be the biggest source of bloat in JAMstack builds. There are solutions out there like Cloudinary that allow you to delegate the image optimization to a service with an easy to use URL-based API for getting optimized images. For users that want to use an approach like this, they don't need to pay the cost of having a larger set of built-in NPM dependencies.

Or users can include whatever image optimization step in their builds that they would like.

The current elm-pages build has a naive, fixed image optimization webpack setup. It was never meant as a long-term solution, and often leads to bulky images that aren't properly compressed, while also slowing down the builds. So removing this from the process will lead to faster builds, and give users more control over that process to optimize images as they wish.

### Extensible favicon/PWA icons instead of a built-in process
There is currently a webpack plugin that is used in the elm-pages build to generate favicons from a single image source.

Again, this adds some bloat to the build step. If a user wanted to use a service like Cloudinary to serve up optimized PWA/Favicon images, the current elm-pages API doesn't allow that because it assumes that it will generate those images for you. So you need to pay the cost of running that on each build, and there's no ability to customize that.

My plan is to give a pure Elm hook that lets you put a set of images to use for your PWA config and as favicons. It will likely use `ImagePath`s, which can either be local images in the `images/` folder (which can be accessed from the `Pages.images` record). Or they can be external URLs, which could point to a service like cloudinary or some other service.

Using the webpack plugin here was a good starting point, but it gives us less control over these things, and has even had some bugs which we don't really have control over (see #146).

## Features that will be added in the process
Besides more control over these things as work that's done directly by elm-pages (not work done by webpack plugins), here are some things that the new elm-pages build will now do:

- Run elm-optimize-level-2 on the final output that's shipped to browsers
- Give users more control over the favicons/PWA icon set they use
- Remove a lot of duplicate processing because elm-pages can pre-render pages in pure Elm, rather than as a 2-step process

**Comment by erlandsona** [comment]
I agree with all of this.
elm-pages to me is about "Doing everything in Elm" that means using elm-ui instead of css, and as little JS as possible.
And the suggestions for other assets like favicons and other images sound great!

**Comment by dillonkearns** [comment]
It will be nice to get rid of these dependencies to speed up the build process once it's not dependent on webpack! That will give a noticeable performance boost to the build process, even without all the other performance improvements.

```diff
diff --git a/package.json b/package.json
index a98799b..c21f197 100644
--- a/package.json
+++ b/package.json
@@ -19,47 +19,17 @@
   "dependencies": {
-    "@babel/core": "^7.5.5",
-    "@babel/preset-env": "^7.5.5",
-    "babel-loader": "^8.0.6",
-    "copy-webpack-plugin": "^5.0.4",
     "cross-spawn": "6.0.5",
-    "css-loader": "^3.2.0",
-    "elm": "^0.19.1-3",
-    "elm-hot-webpack-loader": "^1.1.2",
-    "elm-webpack-loader": "^6.0.0",
     "express": "^4.17.1",
-    "favicons-webpack-plugin": "^3.0.0",
-    "file-loader": "^4.2.0",
-    "find-elm-dependencies": "2.0.2",
     "globby": "^10.0.1",
     "gray-matter": "^4.0.2",
-    "html-webpack-plugin": "^4.2.0",
-    "imagemin-mozjpeg": "^8.0.0",
-    "imagemin-webpack-plugin": "^2.4.2",
-    "lodash": "4.17.15",
-    "node-sass": "^4.12.0",
-    "prerender-spa-plugin": "^3.4.0",
-    "raw-loader": "^4.0.0",
-    "sass-loader": "^8.0.0",
-    "script-ext-html-webpack-plugin": "^2.1.4",
     "sharp": "^0.25.2",
-    "style-loader": "^1.0.0",
-    "temp": "^0.9.0",
-    "terser-webpack-plugin": "^2.3.5",
-    "webpack": "4.42.1",
-    "webpack-dev-middleware": "^3.7.0",
-    "webpack-hot-middleware": "^2.25.0",
-    "webpack-merge": "^4.2.1",
-    "workbox-webpack-plugin": "^4.3.1",
     "xhr2": "^0.2.0"
   },
   "devDependencies": {
     "@types/express": "^4.17.0",
     "@types/jest": "^25.2.2",
     "@types/node": "^12.7.7",
-    "@types/webpack": "^4.32.1",
-    "@types/webpack-dev-middleware": "^2.0.3",
     "elm-format": "^0.8.2",
     "elm-test": "^0.19.1-revision2",
     "jest": "^26.0.1",
@@ -74,4 +44,4 @@
}
```

**Comment by erlandsona** [comment]
This is amazing

**Comment by twop** [comment]
Love it!

**Comment by tennety** [comment]
All of this makes a lot of sense to me, and is very much in keeping with the Elm philosophy of picking the one thing and doing it well. Reflecting on how quickly I was able (as an advanced beginner at Elm) to get up and running with elm-pages, my first thought was that it would get somewhat harder for folks to do so without these defaults. But then I realized that I had used the starter repo and not elm-pages directly; and this approach I think makes it even more possible for folks to make starter templates that "bring their own" stacks of bundlers, pre-processors etc without butting up against the default choices.

![](https://media0.giphy.com/media/1lyivKJRVYnU6E259L/giphy-downsized.gif?cid=6104955e2cdd83c8b09cbd9a5ef4f9d3972109a46fa4cb6b&rid=giphy-downsized.gif)

**Comment by dillonkearns** [comment]
This is now available in the elm-pages NPM package version 1.5.0 with Elm package 7.0.0!

Here are the instructions: https://github.com/dillonkearns/elm-pages/blob/master/docs/7.0.0-elm-package-upgrade-guide.md#2---beta-build-command.

I'm still looking for feedback before removing the current webpack-based `elm-pages build` command and replacing it with the current beta. So anybody who has thoughts or suggestions, I'm all ears!

**Comment by dillonkearns** [comment]
The commit is official now in this branch!

https://github.com/dillonkearns/elm-pages/pull/176/commits/c9c63b8a3b0e1693096ab26a40a7234316ae8fe8#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519

I've got some more features I'd like to get in before the release and I'll do a pre-release beta feedback round. But after that, we'll be seeing a much leaner `elm-pages`! 🎉

**Comment by dillonkearns** [comment]
And it's live on NPM now!

**Comment by TSFoster** [comment]
@dillonkearns I came to try out elm-pages 2 straight after listening to your most recent episode of Elm Radio. What a fantastic release, I'm excited to try it out on a couple of projects that I know it will be a perfect fit for 😃.

I've read this issue and everything you wrote makes total sense. I wonder though what your thoughts are on the files that are inherently related to the data elm-pages uses and creates. Files like sitemaps, RSS feeds, redirects files like [those Netlify supports](https://docs.netlify.com/routing/redirects/). Am I right in thinking they can no longer be generated by elm-pages? Do you think it would be better (in terms of an elm-pages philosophy) to generate them by other means? Or is it just that they had to go in their current incarnation to make the improvements you have?

**Comment by dillonkearns** [comment]
Hello @TSFoster! Sitemaps can be generated using the `getStaticPaths` that's passed into `Api.elm`:

https://github.com/dillonkearns/elm-pages/blob/56dec4752ce538f5a99b7f607ba1098295ff35ca/examples/docs/src/Api.elm#L85-L102

RSS feeds can be generated as well in a similar way. Basically, the way I'm thinking of it is that `elm-pages`' job is data in and data out. That is, get data in from the right place and get it out to the right places (like outputting files). But everything in between (the actual contents) is up to the user.

To give some concrete examples:

`elm-pages` platform:

- Read in file contents
- Read frontmatter from files
- Generate files (Api.elm module)
- Provide a way to pre-render pages

Userland:
- Decide how to decode frontmatter data (or whether to use frontmatter from files at all or something totally different)
- Map files to which pages will be pre-rendered (for example, could be one-to-one mapping of files, or could be multiple routes from one file, or could come from something besides files, etc.)

Hope that clarifies the philosophy a bit!

**Comment by TSFoster** [comment]
Ah I'm so sorry I missed that, I don't know how I managed it. I thought I'd read through the documentation properly but I'd clearly not paid enough attention to the documentation for the elm package. The issues I found on Github were all to do with v1, but I should have just jumped in and figured it out.

Looking through the documentation on elm-pages.com again, I note that on the [File Structure page](https://elm-pages.com/docs/file-structure#api.elm) there's a link at the top to the Api.elm section, but it doesn't exist. Would it be helpful for me to create a new issue about it?
